lib/greeve/base_item.rb

```diff
@@ -0,0 +1,240 @@
+    def self.endpoint(path)
+      # Remove leading slash and file extension.
+      @endpoint = path.gsub(/\A\/*(.*?)(?:\.xml)?(?:\.aspx)?\z/, '\1')
+    end
+
+    # @abstract Subclass and use the DSL methods to map API endpoints to objects
+    #
+    # @option opts [String, Fixnum] :key API key
+    # @option opts [String] :vcode API vCode
+    # @option opts [Hash<String, String>] :query_params a hash of HTTP query
+    #   params that specify how a value maps to the API request
+    #
+    # @example
+    #   super(query_params: {
+    #     "characterID" => character_id,
+    #   })
+    def initialize(opts = {})
+      @api_key = opts[:key]
+      @api_vcode = opts[:vcode]
+      @query_params = opts[:query_params] || {}
+
+      raise TypeError, "Cannot instantiate an abstract class" \
+        if self.class.superclass != Greeve::BaseItem
+
+      refresh
+    end
+
+    # Query the API, refreshing this object's data.
+    #
+    # @return true if the endpoint was fetched (HTTP request sent), false if
+    #   the cache hasn't expired
+    def refresh
+      return false unless cache_expired?
+
+      fetch
+      true
+    end
+
+    # @return true if the API cache timer has expired and this object can
+    #   be refreshed
+    def cache_expired?
+      !(cached_until && cached_until > Time.now)
+    end
+
+    # @return [Time, nil] time when the cache expires and the resource can be
+    #   refreshed (sends an HTTP request)
+    def cached_until
+      return unless @xml_element
+
+      _cached_until = @xml_element.locate("eveapi/cachedUntil/?[0]").first
+      _cached_until ? Time.parse(_cached_until + " UTC") : nil
+    end
+
+    # :nodoc:
+    def inspect
+      attrs = to_s
+
+      unless attrs.empty?
+        attrs = attrs.split("\n").map { |l| "  #{l}" }.join("\n")
+        attrs = "\n#{attrs}\n"
+      end
+
+      "#<#{self.class.name}:#{object_id}#{attrs}>"
+    end
+
+    # @return [String] a string representation of the non-nil attributes
+    def to_s
+      to_h
+        .map { |k, v|
+          v = v.to_s("F") if v.is_a?(BigDecimal)
+          "#{k}: #{v}"
+        }
+        .join("\n")
+    end
+
+    # @return [Hash] a hash of non-nil attributes
+    def to_h
+      attributes
+        .keys
+        .map { |name|
+          value = __send__(name)
+          value = value.to_a if value.is_a?(Rowset)
+
+          [name, value]
+        }
+        .to_h
+        .reject { |k, v| v.nil? }
+    end
+
+    private
+
+    # @return [Hash] the hash of attributes for this object
+    def attributes
+      self.class.instance_variable_get(:@attributes) || {}
+    end
+
+    # @return [String] the class's endpoint path
+    def endpoint
+      self.class.instance_variable_get(:@endpoint)
+    end
+
+    # Fetch data from the API HTTP endpoint.
+    def fetch
+      url = "#{Greeve::EVE_API_BASE_URL}/#{endpoint}.xml.aspx"
+
+      query_params = @query_params
+
+      if @api_key && @api_vcode
+        query_params.merge!({
```

Comment:
The query params need to be duped so it doesn't overwrite the instance variable.